soffice: Report keyboard-triggered formatting toggles in Writer

michaelweghorn opened the pull request:
### Link to issue number:

Fixes #4248

### Summary of the issue:

When toggling text formatting in LibreOffice Writer using keyboard shortcuts (eg. Ctrl+B for toggling bold when using English UI), NVDA was not reporting whether bold was enabled or not.

### Description of user facing changes

In LibreOffice Writer (version 24.8 and newer), when toggling text formatting (bold, italic, underline, subscript/superscript, alignment) using the corresponding keyboard shortcut, NVDA announces the new formatting attribute (e.g. "Bold on", "Bold off"), similar to how it does for Microsoft Word.

### Description of development approach

In the SymphonyDocument class,
register gestures for keyboard shortcuts that
can be used to toggle formatting in LibreOffice Writer.

Add a method to handle the gestures, which is the same for all attributes and enables processing of state change events on toolbar buttons until a timeout
(of currently 0.15 seconds) is reached.
During that time, when a state change event for
a toolbar button is received from LibreOffice,
the name of the button and its new state (on/off)
are announced.

This also requires a corresponding change in LibreOffice [1] so that the state change events for toolbar buttons are actually sent, contained in LibreOffice >= 24.8:

    commit 0425b6eb47830b1fe630dc0128d5049f4b3e5582
    Author: Michael Weghorn <m.weghorn@posteo.de>
    Date:   Tue Apr 16 19:02:30 2024 +0200

        tdf#160695 wina11y: Send status change events for toolbar buttons

Note: An alternative approach querying the text attributes from the IAccessibleText interface instead of the button state would run into the problem that it works reliably when text is selected, but has no clear way to figure out whether or not bold is enabled or not when there's no selection. (The text attributes need to be queried for a given index, but when the text cursor is e.g. in front of the first character of non-bold text, Ctrl+B would enable bold for newly typed text, but querying the text attributes at index 0 of the existing text would still not report bold, resulting in issues like as observed and described in issue #16412 for MS Word.)

[1] https://git.libreoffice.org/core/commit/0425b6eb47830b1fe630dc0128d5049f4b3e5582

### Testing strategy:

1. start NVDA
2. start LibreOffice Writer (current development version, e.g. an upcoming daily build available from [2]).
3. Press Ctrl+B to enable Bold
4. Verify that NVDA announces "Bold on"
5. Press Ctrl+B again to disable bold formatting
6. Verify that NVDA announces "Bold off"
7. Type "hello world"
8. repeat steps 3-6
9. move text cursor to the beginning of the text
10. repeat steps 3-6
11. select the text
12. repeat steps 3-6
13. repeat steps 1-12 for other formatting attributes (Ctrl+I for italic, Ctrl+U for underline,...)

[2] https://dev-builds.libreoffice.org/daily/master/current.html

### Known issues with pull request:

None

### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

michaelweghorn committed:
soffice: Report keyboard-triggered formatting toggles in Writer

 ### Link to issue number:

Fixes #4248

 ### Summary of the issue:

When toggling text formatting in LibreOffice Writer using
keyboard shortcuts (eg. Ctrl+B for toggling bold when using
English UI), NVDA was not reporting whether bold was enabled
or not.

 ### Description of user facing changes

In LibreOffice Writer (version 24.8 and newer), when toggling text
formatting (bold, italic, underline, subscript/superscript, alignment)
using the corresponding keyboard shortcut, NVDA announces
the new formatting attribute (e.g. "Bold on", "Bold off"),
similar to how it does for Microsoft Word.

 ### Description of development approach

In the SymphonyDocument class,
register gestures for keyboard shortcuts that
can be used to toggle formatting in LibreOffice Writer.

Add a method to handle the gestures, which is the same
for all attributes and enables processing of state
change events on toolbar buttons until a timeout
(of currently 0.15 seconds) is reached.
During that time, when a state change event for
a toolbar button is received from LibreOffice,
the name of the button and its new state (on/off)
are announced.

This also requires a corresponding change in LibreOffice [1]
so that the state change events for toolbar buttons
are actually sent, contained in LibreOffice >= 24.8:

    commit 0425b6eb47830b1fe630dc0128d5049f4b3e5582
    Author: Michael Weghorn <m.weghorn@posteo.de>
    Date:   Tue Apr 16 19:02:30 2024 +0200

        tdf#160695 wina11y: Send status change events for toolbar buttons

Note: An alternative approach querying the text attributes
from the IAccessibleText interface instead of the button
state would run into the problem that it works reliably when
text is selected, but has no clear way to figure out whether
or not bold is enabled or not when there's no selection.
(The text attributes need to be queried for a given index, but
when the text cursor is e.g. in front of the first character of
non-bold text, Ctrl+B would enable bold for newly typed text, but
querying the text attributes at index 0 of the existing text
would still not report bold, resulting in issues like
as observed and described in issue #16412 for MS Word.)

[1] https://git.libreoffice.org/core/commit/0425b6eb47830b1fe630dc0128d5049f4b3e5582

 ### Testing strategy:

1. start NVDA
2. start LibreOffice Writer (current development version, e.g. an
   upcoming daily build available from [2]).
3. Press Ctrl+B to enable Bold
4. Verify that NVDA announces "Bold on"
5. Press Ctrl+B again to disable bold formatting
6. Verify that NVDA announces "Bold off"
7. Type "hello world"
8. repeat steps 3-6
9. move text cursor to the beginning of the text
10. repeat steps 3-6
11. select the text
12. repeat steps 3-6
13. repeat steps 1-12 for other formatting attributes
    (Ctrl+I for italic, Ctrl+U for underline,...)

[2] https://dev-builds.libreoffice.org/daily/master/current.html

 ### Known issues with pull request:

None

 ### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

CyrilleB79 commented:
I am not a user of LO and will probably not perform tests.

One question though. Are the toolbars or their buttons always present or is this configurable by the user (e.g. show/hide toolbar, remove button, etc.)?

CyrilleB79 commented:
> In the case of removing buttons or hiding toolbars, formatting toggles are no longer announced. At first thought, this seems even somewhat in line with what sighted users experience: If the buttons are hidden, there is no more visual feedback either.

This makes sense. Just check that NVDA fails gracefully in this case or in the case Libre Office is too old, i.e. nothing reported, but no error in the log; a debugWarning may be suitable though.

> An alternative approach to relying on the buttons I thought of initially was using text attributes, but that cannot reliably detect whether e.g. bold is on/off when there's no text selection, s. PR description.

Yes, same for MS Word.

CyrilleB79 commented on source/appModules/soffice.py:
```diff
@@ -350,6 +352,40 @@ def event_selectionWithIn(self):
+            # button's accessible name is the font attribute, e.g. "Bold", "Italic"
+            if enabled:
+                # Translators: a message when toggling formatting (e.g. bold, italic) in LibreOffice
+                message = _("{textAttribute} on").format(textAttribute=self.name)
```
Is `self.name` translated in non-English UI of LO?

michaelweghorn replied:
Yes. I just tested with German UI, and pressing Ctrl+B causes the announcement of "Blocksatz on".  This is not the equivalent of "Bold", but "Justified", as the German keyboard shortcuts are different.
From what I could see, taking care that the right keyboard shortcuts/gestures are handled is generally a responsibility of the corresponding translation (i.e. language communities). Is that correct?

CyrilleB79 replied:
Yes until now, localization of shortcuts are handled by translators.

This may change or not change in a near future since the translation process is currently being modified.

But until NV Access gives further instructions, let's stick to the old method.

CyrilleB79 commented on source/appModules/soffice.py:
```diff
@@ -433,6 +474,35 @@ def _backspaceScriptHelper(self, unit: str, gesture: inputCore.InputGesture):
+		# send gesture
+		gesture.send()
+
+	__gestures = {
```
The preferred method to associate gestures to scripts is the `@script` decorator; could you use it? Or is there any reason why you did not?

Note that you can define the `@script` decorator with only the `gesture` or the `gestures` parameters, i.e. you can omit `description` and `category` so that the script is not remappable.

michaelweghorn replied:
Thanks, I've updated this accordingly.
(The only reason I didn't do that was that I previously looked at another example that didn't (`script_toggleBold` in `source/NVDAObjects/window/winword.py`) and did the same as that one, rather than starting from scratch using the [NVDA developer guide where this is explained](https://www.nvaccess.org/files/nvda/documentation/developerGuide.html#DefiningScriptProperties).)

CyrilleB79 reviewed (commented):
Below are some comments on the code.

Also, did you test with quick key typing? E.g. double press ``control+b` or even keep `control+b` pressed during two seconds.

michaelweghorn committed:
Follow-up: Use @script decorator

As suggested in review:
https://github.com/nvaccess/nvda/pull/16413#discussion_r1568980382

seanbudd commented on user_docs/en/changes.t2t:
```diff
@@ -44,6 +44,7 @@ What's New in NVDA
   - The volume of the other applications can be adjusted by ``NVDA+alt+pageUp`` and ``NVDA+alt+pageDown``. (#16052, @mltony)
   - The sound of the other applications can be muted with ``NVDA+alt+delete``. (#16052, @mltony)
   -
+- In LibreOffice Writer (version 24.8 and newer), when toggling text formatting (bold, italic, underline, subscript/superscript, alignment) using the corresponding keyboard shortcut, NVDA announces the new formatting attribute (e.g. "Bold on", "Bold off"). (#4248, @michaelweghorn)
```
this will need to be moved to 2024.3 when the change log section is created

michaelweghorn replied:
Is there anything to do for me at the moment or just wait for now?

seanbudd replied:
just wait

seanbudd commented on source/appModules/soffice.py:
```diff
@@ -350,6 +353,40 @@ def event_selectionWithIn(self):
 
+class SymphonyButton(IAccessible):
+
+    def event_stateChange(self):
```
please add a return type

seanbudd commented on source/appModules/soffice.py:
```diff
@@ -399,6 +436,11 @@ def _get_locationText(self):
 
+	# variables used for handling announcements resulting from gestures
+	GESTURE_ANNOUNCEMENT_TIMEOUT = 0.15
+	announce_toolbar_button_toggle = False
+	last_formatting_gesture_event_time = 0
```
please use lowerCamelCase for variables

michaelweghorn replied:
done

seanbudd commented on source/appModules/soffice.py:
```diff
@@ -433,6 +475,36 @@ def _backspaceScriptHelper(self, unit: str, gesture: inputCore.InputGesture):
+			"kb:control+j",
+		]
+	)
+	def script_toggleTextAttribute(self, gesture):
```
please add a type hint to gesture

michaelweghorn replied:
done

seanbudd commented on source/appModules/soffice.py:
```diff
@@ -433,6 +475,36 @@ def _backspaceScriptHelper(self, unit: str, gesture: inputCore.InputGesture):
+	)
+	def script_toggleTextAttribute(self, gesture):
+		# reset time and enable announcement of toggled toolbar buttons
+		# s. SymphonyButton#event_stateChange
```
typo here

michaelweghorn replied:
Changed to this now:

    # see L{SymphonyButton.event_stateChange}

Is this the right way to reference that method? I've seen that being used elsewhere.

(From a quick search for Sphinx examples, I've found mentions of syntax using `:func:<method>`, which would potentially mean

    # see :func:`SymphonyButton.event_stateChange`

here? But I didn't see such syntax being used anywhere in the NVDA code base, so used what I saw elsewhere for now.)

seanbudd replied:
We have only just migrated to using Sphinx for new code instead of epydoc, I think using the Sphinx method is preferred here

michaelweghorn replied:
OK, switched to the Sphinx method (as I understand it) now.

seanbudd reviewed (commented):
Thanks @michaelweghorn , just minor issues

michaelweghorn committed:
Follow-up: Address review comments

s. https://github.com/nvaccess/nvda/pull/16413#pullrequestreview-2007798779

CyrilleB79 commented on source/appModules/soffice.py:
```diff
@@ -433,6 +475,36 @@ def _backspaceScriptHelper(self, unit: str, gesture: inputCore.InputGesture):
+		]
+	)
+	def script_toggleTextAttribute(self, gesture: inputCore.InputGesture):
+		# reset time and enable announcement of toggled toolbar buttons
+		# see L{SymphonyButton.event_stateChange}
```
Now that you are using `@script` decorator, you can safely use the docstring for this function instead of comments.

```suggestion
		"""Reset time and enable announcement of toggled toolbar buttons.
		See L{SymphonyButton.event_stateChange}
		"""
```

michaelweghorn replied:
Thanks!

michaelweghorn committed:
Follow-up: Use function docstring instead of comments

s. https://github.com/nvaccess/nvda/pull/16413#pullrequestreview-2008231160

Co-authored-by: Cyrille Bougot <cyrille.bougot2@laposte.net>

michaelweghorn committed:
Follow-up: Use Sphinx syntax for reference

s. https://github.com/nvaccess/nvda/pull/16413#discussion_r1571759738

seanbudd committed:
Merge remote-tracking branch 'origin/master' into michaelweghorn/soffice_report_formatting_toggles

CyrilleB79 commented:
These shortcuts need to be translated by translators for non-English GUIs. In French for example, it seems that all shortcuts are the same except for bold which is control+G instead of control+B.

@michaelweghorn, I have tested with a portable version of LO < 24.8 (maybe 24.2.4). Strangely, it was already working when toggling "underline" but not with the other ones. Any explanation?

More generally, would you be able to point me in LO's code where the localized formatting shortcuts are defined. This may be a helpful information to pass to translators.
Thanks.

michaelweghorn commented:
@CyrilleB79:  Thanks for looking into this.

> @michaelweghorn, I have tested with a portable version of LO < 24.8 (maybe 24.2.4). Strangely, it was already working when toggling "underline" but not with the other ones. Any explanation?

Before LibreOffice commit https://git.libreoffice.org/core/commit/0425b6eb47830b1fe630dc0128d5049f4b3e5582 , toolbar items that have a "toggle button" role in LO wouldn't send the necessary event. However, the "underline" control in the formatting toolbar has a "dropdown button" role, which is presumably why it was unaffected and already working fine.

> More generally, would you be able to point me in LO's code where the localized formatting shortcuts are defined. This may be a helpful information to pass to translators.

They're defined in this XML file: https://git.libreoffice.org/core/+/refs/heads/master/officecfg/registry/data/org/openoffice/Office/Accelerators.xcu

For the Ctrl+G example in French you mention, that would be the this entry:
https://git.libreoffice.org/core/+/refs/heads/master/officecfg/registry/data/org/openoffice/Office/Accelerators.xcu#1850

```
        <node oor:name="G_MOD1" oor:op="replace">
          <prop oor:name="Command">
            <value xml:lang="x-no-translate">L10N SHORTCUTS - NO TRANSLATE</value>
            <value xml:lang="es">.uno:Save</value>
            <value xml:lang="fr">.uno:Bold</value>
          </prop>
        </node>
```
`G_MOD1` means: G + the `MOD1` key, where `MOD1` is Ctrl (on Windows and Linux at least), so that's G+Ctrl.

For German, bold (`.uno:Bold`) would be this entry:

```
        <node oor:name="F_SHIFT_MOD1" oor:op="replace">
          <prop oor:name="Command">
            <value xml:lang="x-no-translate">L10N SHORTCUTS - NO TRANSLATE</value>
            <value xml:lang="de">.uno:Bold</value>
          </prop>
        </node>
```
where `F_SHIFT_MOD1` means F + Shift + MOD1 (= F+Shift+Ctrl)

The relevant so-called "UNO commands" to look for are:

* `.uno:Bold`
* `.uno:Italic`
* `.uno:Underline`
* `.uno:SuperScript`
* `.uno:SubScript`
* `.uno:LeftPara` (align left)
* `.uno:CenterPara` (align center)
* `.uno:RightPara` (align right),
* `.uno:JustifyPara` (justified)

If there's a language-specific entry, like `<value xml:lang="fr">.uno:Bold</value>` for French, that shortcut will be used, otherwise the default "en-US" one will be used, e.g. `<value xml:lang="en-US">.uno:Bold</value>` in the case of bold).

CyrilleB79 commented:
Thanks for the details. I'll send make the chenges for French and send a message on the translators mailing list.
